### PR TITLE
feat(etcdv3): support unsubscribe and subscription recovery

### LINKED ISSUE #3366: [FEATURE] etcdv3 registry: implement DoUnregister, DoUnsubscribe and reconnect subscription recovery
### ⚠️ Verification

- [x] I have searched the [issues](https://github.com/apache/dubbo-go/issues) of this repository and believe that this is not a duplicate.
- [x] I have searched the [release notes](https://github.com/apache/dubbo-go/releases) of this repository and believe that this is not a duplicate.

### 🎯 Solution Description

The etcdv3 registry had three critical gaps compared to the zookeeper registry:

1. **DoUnregister was not implemented** — always returned `"DoUnregister is not support in etcdV3Registry"`.
   The fix deletes the corresponding etcd key via `client.Delete`, guarded by a `client.Valid()` check.

2. **DoUnsubscribe was not implemented** — always returned an error.
   The fix introduces a per-service listener architecture: `dataListener` now holds a
   `subscribed map[string]ConfigurationListener` keyed by service key, replacing the old
   single shared `configurationListener`. Each `DoSubscribe` call creates an independent
   `configurationListener` for that service; `DoUnsubscribe` closes and removes it.

3. **Subscriptions were silently lost after reconnect** — `InitListeners` (called by `RestartCallBack`
   on reconnect) replaced `dataListener` with an empty new instance, discarding all active
   subscriptions. The fix stores `subscribeURL *common.URL` on each `configurationListener` and
   adds a recovery loop in `InitListeners` that mirrors the zookeeper registry: close stale
   listeners, create fresh ones, restart `ListenServiceEvent` goroutines for every recovered URL.

etcdv3 注册中心相比 zookeeper 注册中心存在三个关键缺口：

1. **DoUnregister 未实现** — 始终返回 `"DoUnregister is not support in etcdV3Registry"`。
   修复：通过 `client.Delete` 删除对应 etcd key，调用前做 `client.Valid()` 校验。

2. **DoUnsubscribe 未实现** — 始终返回错误。
   修复：引入 per-service listener 架构，`dataListener` 改为以 service key 为键的
   `subscribed map`，取代原来全局共享的单个 `configurationListener`。每次 `DoSubscribe`
   为该服务创建独立的 `configurationListener`；`DoUnsubscribe` 关闭并移除对应 listener。

3. **断线重连后订阅静默丢失** — `InitListeners`（重连时由 `RestartCallBack` 触发）将
   `dataListener` 替换为空实例，所有活跃订阅被丢弃。
   修复：在 `configurationListener` 上存储 `subscribeURL *common.URL`，在 `InitListeners`
   中加入与 zookeeper 注册中心对齐的恢复循环：关闭旧 listener，为每个已订阅 URL 创建新
   listener，重新启动 `ListenServiceEvent` goroutine。

### 📋 Use Cases

- Services that need to deregister from etcd on graceful shutdown.
- Services that dynamically subscribe/unsubscribe to providers at runtime.
- Long-running applications deployed in environments with unstable etcd connections (container
  restarts, network partitions, etcd leader elections) that require seamless reconnect without
  losing active subscriptions.

- 需要在优雅停机时从 etcd 注销的服务。
- 运行时动态订阅/取消订阅 provider 的服务。
- 部署在网络不稳定环境（容器重启、网络分区、etcd leader 选举）中、要求无感知重连且不丢失
  订阅的长期运行应用。

### ⚖️ Complexity & Risks

- Medium complexity. The per-service listener refactor touches the core subscribe/unsubscribe path.
- `DataChange` now iterates all subscribed keys (instead of returning on first match), which is
  a correctness improvement for wildcard/any-condition scenarios but changes the semantics slightly.
- `InitListeners` holds the old `dataListener` mutex for the duration of recovery via `defer`,
  which may briefly delay concurrent `DataChange` calls — same trade-off as zookeeper.

- 中等复杂度，per-service listener 重构影响核心订阅/取消订阅路径。
- `DataChange` 改为遍历全部匹配 key（原先匹配即返回），对 wildcard/any-condition 场景是正确性
  修复，但语义有轻微变化。
- `InitListeners` 通过 `defer` 持有旧 `dataListener` mutex 直至恢复完成，大量订阅时可能短暂
  阻塞并发 `DataChange`，与 zookeeper 注册中心权衡一致。

### 🔗 External Dependencies

None. Changes are confined to `registry/etcdv3/listener.go` and `registry/etcdv3/registry.go`,
using existing `remoting/etcdv3` and `common.URL` APIs.

无外部依赖，改动限于 `registry/etcdv3/listener.go` 和 `registry/etcdv3/registry.go`，
使用已有的 `remoting/etcdv3` 和 `common.URL` API。

### 📘 Additional Context

Reference implementation for reconnect recovery: `registry/zookeeper/registry.go` — `InitListeners` (line 92-120).

重连恢复逻辑参考实现：`registry/zookeeper/registry.go` — `InitListeners`（第 92-120 行）。

如果觉得可以的话Assign me 然后我了解到我们目前需要将新注册中心迁移到dubbo-go-extension内作为插件使用 可以实现后进行迁移

### PR DESCRIPTION
**Pull request by wyf027**
### Description

Fixes #3366.

This PR fills the missing etcdv3 registry subscription lifecycle behavior:

- implements `DoUnregister` by deleting the registered etcd key after validating the client
- replaces the single shared config listener with per-service subscribed listeners
- implements `DoUnsubscribe` by closing and removing the service listener
- preserves active subscriptions when listeners are reinitialized after client restart
- dispatches data changes to every matching subscribed service key

### Validation

- `go test ./registry/etcdv3`
- `gofmt` on touched files
- `git diff --check`

### Checklist

- [x] Code has passed local testing
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by wyf027** [comment]
Addressed the remaining Codecov patch coverage feedback by covering the delete-event skip path and the valid-client DoUnregister path. Local validation passes: go test -coverprofile=/tmp/etcdv3.cover ./registry/etcdv3, go test ./registry/etcdv3 ./remoting/etcdv3, and git diff --check.

File: registry/etcdv3/listener.go [hunk 1]
```diff
@@ -37,18 +37,37 @@ import (
 )
 
 type dataListener struct {
-	interestedURL []*common.URL
-	listener      config_center.ConfigurationListener
+	subscribed map[string]config_center.ConfigurationListener
+	mutex      sync.Mutex
```
Comment (AlexStocks) [inline registry/etcdv3/listener.go:40]: 这个 mutex是为了guard subscribed 的吗？如是，挪到 subscribed  上面去

**Comment by wyf027** [comment]
Follow-up for the remaining Codecov partial coverage warning in registry/etcdv3/listener.go: added a focused test that drives configurationListener.Next through the skipped delete-event path when the etcd client is invalid, then verifies the next add event is returned. Local validation passes: go test ./registry/etcdv3 ./remoting/etcdv3, go test -race ./registry/etcdv3, go test -coverprofile=/tmp/etcdv3_after.out ./registry/etcdv3, and git diff --check.

File: registry/etcdv3/registry.go [hunk 2]
```diff
@@ -158,14 +195,41 @@ func (r *etcdV3Registry) DoSubscribe(svc *common.URL) (registry.Listener, error)
...
+	r.dataListener.mutex.Lock()
+	subscribedListener := r.dataListener.subscribed[conf.ServiceKey()]
+	if subscribedListener != nil {
+		etcdListener, _ := subscribedListener.(*configurationListener)
+		if etcdListener != nil && etcdListener.closed() {
+			r.dataListener.mutex.Unlock()
+			return nil, perrors.Errorf("configListener for service %s has already been closed", conf.ServiceKey())
+		}
+	}
+	listener := r.dataListener.unsubscribeURLLocked(conf)
```
Comment (AlexStocks) [inline registry/etcdv3/registry.go]: [P1] 这里在校验 `r.listener == nil` 之前已经调用 `unsubscribeURLLocked`，该调用会关闭 listener 并从 `subscribed` 中删除记录。若下面返回 `etcd event listener is nil`，`BaseRegistry.UnSubscribe` 会把本次退订视为失败，但本地订阅已经被永久移除，后续恢复或重试都拿不到原 listener。请先完成 `r.listener`/类型校验，再执行关闭和删除，或在错误路径回滚订阅 map。

**Comment by wyf027** [comment]
Addressed the P1 DoUnsubscribe feedback. DoUnsubscribe now validates the event listener and the subscribed entry type before calling unsubscribeURLLocked, so error paths no longer close/remove the local subscription. I also added a regression test covering the nil event-listener path and verifying the subscription remains in the map and open. Local validation passes: go test ./registry/etcdv3 ./remoting/etcdv3, go test -race ./registry/etcdv3, go test -coverprofile=/tmp/etcdv3_fix.out ./registry/etcdv3, and git diff --check.

**Comment by wyf027** [comment]
Follow-up for the failed CI lint job: changed the new error assertion to require.ErrorContains to satisfy testifylint. Local validation passes: go test ./registry/etcdv3 ./remoting/etcdv3, go test -race ./registry/etcdv3, and git diff --check. golangci-lint is not installed in this local environment, so I verified the exact reported lint issue by matching the CI log.
